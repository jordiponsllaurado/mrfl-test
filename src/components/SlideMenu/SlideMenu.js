import { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { StyledMenu } from './SlideMenu.styled';

const SlideMenu = ({ isOpen, sections, sectionActive, onClickOutside }) => {
  const wrapperRef = useRef();

  const renderItem = item => {
    const className = `section ${sectionActive === item ? 'active' : ''}`;
    return (
      <a href="/" className={className} key={item}>
        {item}
      </a>
    );
  };

  const handleClickOutside = event => {
    if (wrapperRef && !wrapperRef.current.contains(event.target)) {
      onClickOutside();
    }
  };

  useEffect(() => {
    if (isOpen) {
      document.addEventListener('mousedown', handleClickOutside);
    } else {
      document.removeEventListener('mousedown', handleClickOutside);
    }
  }, [isOpen]);

  return (
    <StyledMenu open={isOpen} className="slide-menu" ref={wrapperRef}>
      <div className="slide-header">
        <img
          src="https://www.marfeel.com/public/images/landing/header/logo-marfeel.svg"
          alt="Marfeel"
        />
      </div>
      <div className="sections">{sections.map(renderItem)}</div>
    </StyledMenu>
  );
};

SlideMenu.propTypes = {
  isOpen: PropTypes.bool,
  sections: PropTypes.arrayOf(PropTypes.string).isRequired,
  sectionActive: PropTypes.string.isRequired,
  onClickOutside: PropTypes.func
};

SlideMenu.defaultProps = {
  isOpen: false,
  onClickOutside: null
};
export default SlideMenu;
