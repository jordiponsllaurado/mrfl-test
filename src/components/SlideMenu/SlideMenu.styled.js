/* eslint-disable import/prefer-default-export */
import styled from 'styled-components';

export const StyledMenu = styled.nav`
  position: fixed;
  justify-content: center;
  background-color: white;
  transform: ${({ open }) => (open ? 'translateX(0)' : 'translateX(-100%)')};
  height: 100%;
  width: 75%;
  text-align: left;
  top: 0;
  left: 0;
  transition: transform 0.3s ease-in-out;
  overflow-y: auto;

  .slide-header {
    background-color: #4b4b4b;

    img {
      padding: 2rem;
    }
  }

  .sections {
    display: flex;
    flex-direction: column;
    justify-content: center;

    a {
      font-size: 2rem;
      text-transform: uppercase;
      padding: 2rem 0;
      font-weight: bold;
      letter-spacing: 0.5rem;
      color: grey;
      text-decoration: none;
      transition: color 0.3s linear;
      padding: 2rem;
    }

    a.active {
      color: orange;
    }

    .section:not(:last-child) {
      border-bottom: 1px solid #4b4b4b;
    }
  }
`;
