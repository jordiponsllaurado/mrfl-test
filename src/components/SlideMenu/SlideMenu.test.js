import { render } from '@testing-library/react';
import SlideMenu from './SlideMenu';

const sections = ['facilitate', 'administration', 'invoice', 'bandwidth'];

test('renders SlideMenu open', () => {
  const { container } = render(
    <SlideMenu sections={sections} sectionActive="administration" open />
  );
  expect(container).toMatchSnapshot();
});

test('renders SlideMenu close', () => {
  const { container } = render(<SlideMenu sections={sections} sectionActive="administration" />);
  expect(container).toMatchSnapshot();
});
