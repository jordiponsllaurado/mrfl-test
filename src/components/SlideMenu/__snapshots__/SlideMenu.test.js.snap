// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SlideMenu close 1`] = `
<div>
  <nav
    class="sc-bdfBwQ kwmMNf slide-menu"
  >
    <div
      class="slide-header"
    >
      <img
        alt="Marfeel"
        src="https://www.marfeel.com/public/images/landing/header/logo-marfeel.svg"
      />
    </div>
    <div
      class="sections"
    >
      <a
        class="section "
        href="/"
      >
        facilitate
      </a>
      <a
        class="section active"
        href="/"
      >
        administration
      </a>
      <a
        class="section "
        href="/"
      >
        invoice
      </a>
      <a
        class="section "
        href="/"
      >
        bandwidth
      </a>
    </div>
  </nav>
</div>
`;

exports[`renders SlideMenu open 1`] = `
<div>
  <nav
    class="sc-bdfBwQ kwmMNf slide-menu"
  >
    <div
      class="slide-header"
    >
      <img
        alt="Marfeel"
        src="https://www.marfeel.com/public/images/landing/header/logo-marfeel.svg"
      />
    </div>
    <div
      class="sections"
    >
      <a
        class="section "
        href="/"
      >
        facilitate
      </a>
      <a
        class="section active"
        href="/"
      >
        administration
      </a>
      <a
        class="section "
        href="/"
      >
        invoice
      </a>
      <a
        class="section "
        href="/"
      >
        bandwidth
      </a>
    </div>
  </nav>
</div>
`;
